### PR TITLE
build: optimizing patient linking

### PR DESCRIPTION
**Pull request by Orta21**
Part of cs-2038

Issues:

- https://linear.app/metriport/issue/CS-2038

### Dependencies

- Upstream: none
- Downstream: none

### Description

- Optimizing patient linking to properly validate links coming in

### Testing

- Local
  - [x] Local script going through list of links
  - [x] unit tests
- Staging
  - [ ] Run PD to make sure process isnt broken
- Production
  - [ ] Monitor production to ensure links are being flagged accordingly

### Release Plan

- [ ] Merge this

## Summary by CodeRabbit

* **New Features**
  * Patient-link scoring pipeline with cross-validation that can promote matches via shared contact or address.
  * New matching: name, DOB, gender, address, contact, and SSN scoring plus fuzzy string matching.

* **Bug Fixes**
  * Stricter validation: partial matches (e.g., DOB+address or gender+name only) now reject.
  * Telecom normalization: phone/email normalized first; unnormalizable values ignored.

* **Refactor**
  * Batch validation of external links for consistent, efficient processing.

* **Tests**
  * Extensive unit tests covering thresholds, edge cases, and cross-validation.

### CODE REVIEW

File: packages/core/src/mpi/filter-patients/match.ssn.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  const ssn2 = externalPatient.personalIdentifiers
+    ?.filter(id => id.type === "ssn")
+    .map(id => id.value);
+
+  if (ssn1?.length && ssn2?.length) {
+    const ssnMatch = ssn1.some(s1 => ssn2.includes(s1));
+    if (ssnMatch) {
+      return 5;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/core/src/mpi/filter-patients/match.ssn.ts]: Addressed by using flatmap

File: packages/core/src/mpi/filter-patients/match-contact.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  externalPatient: PatientData
+): number {
+  const directMatch = metriportPatient.contact?.some(c1 =>
+    externalPatient.contact?.some(c2 => c1[field] && c2[field] && c1[field] === c2[field])
+  );
+
+  if (directMatch) return 2;
+
+  return 0;
+}
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/core/src/mpi/filter-patients/match-contact.ts]: Its normalized upstream

File: packages/api/src/external/carequality/shared.ts [hunk 3]
```diff
@@ -206,10 +211,15 @@ export function cqLinkToPatientData(cqLink: CQLink): PatientData {
...
+
+      const normalizedPhone = normalizePhoneNumberSafe(value);
+      if (normalizedPhone) {
+        telecom.push({ phone: normalizedPhone });
+      } else {
+        const normalizedEmail = normalizeEmailNewSafe(value);
+        if (normalizedEmail) {
+          telecom.push({ email: normalizedEmail });
+        }
       }
```
Comment (lucasdellabella) [inline packages/api/src/external/carequality/shared.ts]: nit; imo this is easier to read

```
      const normalizedPhone = normalizePhoneNumberSafe(value);
      const normalizedEmail = normalizeEmailNewSafe(value);
      if (normalizedPhone) {
        telecom.push({ phone: normalizedPhone });
      } else if (normalizedEmail) {
        telecom.push({ email: normalizedEmail });
      }
```

File: packages/api/src/external/hie/validate-patient-links.ts
Comment (lucasdellabella) [inline packages/api/src/external/hie/validate-patient-links.ts:1]: The cq + cw implementations are basically identical except that they each use their own __LinkToPatientData function, let's create a shared function that accepts cqLinkToPatientData / cwLinkToPatientData as an argument to reduce duplication.

File: packages/api/src/external/hie/validate-patient-links.ts [hunk 4]
```diff
@@ -38,38 +41,67 @@ export async function validateCwLinksBelongToPatient(
...
-  }
+  const invalidLinks = linkToPatientDataMap
+    .filter(item => invalidPatientDataLinks.includes(item.patientData))
+    .map(item => item.link);
 
   return { validNetworkLinks, invalidLinks };
 }
 
-const validateLinkBelongsToPatient = async (
+async function validateLinksBelongToPatient(
```
Comment (lucasdellabella) [inline packages/api/src/external/hie/validate-patient-links.ts:64]: these are mutually exclusive and the parent ultimately is deciding what is actually a valid link / invalid link.

Can we return only invalid links here and then use partition in the caller to minimize the diff variables we have floating around? More variables = more things for readers to untangle / follow around the callstack.

```
  const [invalidItems, validItems] = partition(
    linkToPatientDataMap,
    item => invalidPatientDataLinks.includes(item.patientData)
  );
```

File: packages/core/src/mpi/filter-patients/filter-patients.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      });
+    } else {
+      invalidLinks.push({
+        patient: patientLink,
+        isMatch,
+        totalScore,
+        scores,
+        failedRule,
+      });
+    }
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/filter-patients.ts]: can use map and lodash's partition to avoid mutability. Also avoids a lot of boilerplatey stuff?

```
const linkStatuses = patientLinks.map(patientLink => ({
  patient: patientLink,
  ...linkFilteringAlgorithm(metriportPatient, patientLink, threshold)
}));

const [validLinks, invalidLinks] = partition(linkStatuses, 'isMatch');
```

File: packages/core/src/mpi/filter-patients/filter-patients.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      failedRule: undefined,
+    }));
+
+  validLinks = [...validLinks, ...crossValidatedLinkStatuses];
+  invalidLinks = invalidLinks.filter(link => !crossValidatedLinks.includes(link.patient));
+
+  return [...validLinks, ...invalidLinks];
+}
+
+export function linkFilteringAlgorithm(
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/filter-patients.ts]: can we rename this to something a little more obvious? "evaluatePatientMatch" or something

File: packages/core/src/mpi/filter-patients/cross-validate-links.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  for (const invalidLink of invalidLinks) {
+    const isValidated = validateLinkThroughValidLinks(invalidLink, validLinks);
+
+    if (isValidated) {
+      crossValidatedLinks.push(invalidLink);
+    }
+  }
+
+  return crossValidatedLinks;
+}
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/cross-validate-links.ts]: I think this can be replaced by a single filter call right?

```
export function crossValidateInvalidLinks(
  validLinks: PatientData[],
  invalidLinks: PatientData[]
): PatientData[] {
  return invalidLinks.filter((invalidLink) => validateLinkThroughValidLinks(invalidLink, validLinks))
}
```

invalidLinks.filter((invalidLink) => validateLinkThroughValidLinks(invalidLink, validLinks))

File: packages/core/src/mpi/filter-patients/cross-validate-links.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+function validateLinkThroughValidLinks(
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/cross-validate-links.ts]: rename to "isLinkActuallyValid" or something since it returns a boolean

File: packages/core/src/mpi/filter-patients/cross-validate-links.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    const hasAddressMatch = invalidLink.address?.some(invalidAddr =>
+      validLink.address?.some(validAddr => isAddressMatch(invalidAddr, validAddr))
+    );
+
+    if (hasAddressMatch) {
+      return true;
+    }
+  }
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/cross-validate-links.ts]: nit; 

anytime i see doubly nested somes i try to see whether i can replace it with a standard set intersection, just a utility where you pass two collections and something that can compare items, and it checks whether there is overlap.

```
    const hasContactMatch = invalidLink.contact && validLink.contact &&
      intersectionWith(invalidLink.contact, validLink.contact, isContactMatch).length > 0;

    if (hasContactMatch) {
      return true;
    }

    const hasAddressMatch = invalidLink.address && validLink.address &&
      intersectionWith(invalidLink.address, validLink.address, isAddressMatch).length > 0;

    if (hasAddressMatch) {
      return true;
    }

File: packages/core/src/mpi/filter-patients/match-address.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+ */
+export function isAddressMatch(addr1: Address, addr2: Address): boolean {
+  return (
+    normalizeString(addr1?.city || "") === normalizeString(addr2?.city || "") &&
+    normalizeString(addr1?.state || "") === normalizeString(addr2?.state || "") &&
+    normalizeString(addr1?.zip || "") === normalizeString(addr2?.zip || "") &&
+    normalizeAddressString(addr1?.addressLine1 || "") ===
+      normalizeAddressString(addr2?.addressLine1 || "")
+  );
+}
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-address.ts]: these two conditions look different - hasAddressMatch doesn't check city. Can we reuse isAddressMatch within hasAddressMatch or are they for different purposes?

File: packages/core/src/mpi/filter-patients/match-address.ts [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+  const directions: { [key: string]: string } = {
+    n: "north",
+    s: "south",
+    e: "east",
+    w: "west",
+    ne: "northeast",
+    nw: "northwest",
+    se: "southeast",
+    sw: "southwest",
+  };
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-address.ts]: pull these out to module level constants as they don't have any dependencies on anything inside of this normalize function.

File: packages/core/src/mpi/filter-patients/match-address.ts [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+      return isFuzzyMatch(addr1Normalized, addr2Normalized, 0.7);
+    })
+  );
+
+  if (cityMatch) score += 0.5;
+  if (zipMatch) score += 0.5;
+  if (stateMatch) score += 0.5;
+  if (addressLine1Match) score += 0.5;
+
+  return score;
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-address.ts]: more stuff that gets made less error prone by using intersectionWith

File: packages/core/src/mpi/filter-patients/match-business-rules.ts [hunk 13]
```diff
@@ -0,0 +1,149 @@
+import { buildDayjs } from "@metriport/shared/common/date";
+import { PatientData } from "../../domain/patient";
+import { calculateLastNameScore } from "./match-name";
+
+export function checkBusinessRules(
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-business-rules.ts:5]: make the code self doucmenting - give a lot of the conditions a name and then use those in your if conditions.

const noNameMatch = scores.names === 0
const noAddressMatch = scores.address < 2
const lowDobMatch = scores.dob <= 1
const noDobMatch = scores.dob === 0
const lastNameScore = calculateLastNameScore(metriportPatient, externalPatient)
...

File: packages/core/src/mpi/filter-patients/match-business-rules.ts [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+  externalPatient: PatientData,
+  scores: {
+    dob: number;
+    gender: number;
+    names: number;
+    address: number;
+    phone: number;
+    email: number;
+    ssn: number;
+  }
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-business-rules.ts]: should be able to simplify this type. `scores: { address: number }` as we don't _require_ any of these other fields.

File: packages/core/src/mpi/filter-patients/match-dob.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    let matchingParts = 0;
+    if (dob1Split[0] && dob2Split[0] && dob1Split[0] === dob2Split[0]) matchingParts++; // Year
+    if (dob1Split[1] && dob2Split[1] && dob1Split[1] === dob2Split[1]) matchingParts++; // Month
+    if (dob1Split[2] && dob2Split[2] && dob1Split[2] === dob2Split[2]) matchingParts++; // Day
+
+    if (matchingParts >= 2) {
+      return 2;
+    } else if (matchingParts === 1) {
+      return 1;
+    }
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-dob.ts]: return Math.ceil(matchingParts, 2) and then you can unnest the else + drop the return 0 at the bottom

File: packages/core/src/mpi/filter-patients/match-gender.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  externalPatient: PatientData
+): number {
+  if (metriportPatient.genderAtBirth && externalPatient.genderAtBirth) {
+    if (metriportPatient.genderAtBirth === externalPatient.genderAtBirth) {
+      return 1;
+    }
+  }
+
+  return 0;
+}
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-gender.ts]: avoid nesting if unnecessary.

```
  return metriportPatient.genderAtBirth 
  && externalPatient.genderAtBirth 
  && metriportPatient.genderAtBirth === externalPatient.genderAtBirth) 
  ? 1
   : 0
```

File: packages/core/src/mpi/filter-patients/match-name.ts [hunk 17]
```diff
@@ -0,0 +1,109 @@
+import { splitName } from "../normalize-patient";
+import { PatientData } from "../../domain/patient";
+import { isFuzzyMatch } from "./utils";
+
+export function calculateNameScore(
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-name.ts]: maybe calculateFullNameScore to minimize potential confusion

File: packages/core/src/mpi/filter-patients/match-name.ts [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+  metriportPatient: PatientData,
+  externalPatient: PatientData
+): number {
+  return (
+    calculateFirstNameScore(metriportPatient, externalPatient) +
+    calculateLastNameScore(metriportPatient, externalPatient)
+  );
+}
+
+export function calculateFirstNameScore(
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-name.ts:15]: For these firstName + lastName score, a bunch of these checks are the same. Could we write it to accept two names and then call it from both places?

File: packages/core/src/mpi/filter-patients/match-business-rules.ts (hunk 13 above)
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-business-rules.ts:5]: this function _always_ returns isMatch: false, totalScore: 0, and score.

Think we should pull all of this out to the callsite, and just return a single failedRule value to minimize boilerplate.

File: packages/core/src/mpi/filter-patients/match-business-rules.ts [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+  dob: number;
+  gender: number;
+  names: number;
+  address: number;
+  phone: number;
+  email: number;
+  ssn: number;
+}): boolean {
+  return scores.phone === 0 && scores.email === 0;
+}
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-business-rules.ts]: Can simplify the typing here, as it only requires two values to be on score

```
function hasNoContact(scores: {
  phone: number;
  email: number;
}): boolean {
  return scores.phone === 0 && scores.email === 0;
}
```

File: packages/core/src/mpi/filter-patients/match-contact.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+  metriportPatient: PatientData,
+  externalPatient: PatientData
+): number {
+  const directMatch = metriportPatient.contact?.some(c1 =>
+    externalPatient.contact?.some(c2 => c1[field] && c2[field] && c1[field] === c2[field])
+  );
+
+  if (directMatch) return 2;
+
+  return 0;
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/match-contact.ts]: return directMatch ? 2 : 0

File: packages/core/src/mpi/filter-patients/filter-patients.ts [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+  metriportPatient: PatientData,
+  externalPatient: PatientData,
+  threshold: number
```
Comment (lucasdellabella) [inline packages/core/src/mpi/filter-patients/filter-patients.ts]: I see that we always set the threshold too 8.5, can we set that as the default value here instead of setting it a dozen times in tests?

File: packages/core/src/mpi/filter-patients/match-address.ts (hunk 10 above)
Comment (Orta21) [reply packages/core/src/mpi/filter-patients/match-address.ts]: Good call i removed isAddressMatch

File: packages/core/src/mpi/filter-patients/match-address.ts [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+      metriportPatient.address,
+      externalPatient.address,
+      (addr1, addr2) => normalizeString(addr1?.zip || "") === normalizeString(addr2?.zip || "")
+    ).length > 0;
+  const stateMatch =
+    intersectionWith(
+      metriportPatient.address,
+      externalPatient.address,
+      (addr1, addr2) => normalizeString(addr1?.state || "") === normalizeString(addr2?.state || "")
+    ).length > 0;
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/core/src/mpi/filter-patients/match-address.ts:35]: not necessary

File: packages/core/src/mpi/filter-patients/match-business-rules.ts [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+  scores: {
+    dob: number;
+    gender: number;
+    names: number;
+    address: number;
+    phone: number;
+    email: number;
+    ssn: number;
+  }
+): string | null {
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/core/src/mpi/filter-patients/match-business-rules.ts:17]: going to leave as is
